<a href="https://colab.research.google.com/github/PrzemyslawSarnacki/NumericalMethods/blob/master/Trygonometric.ipynb" target="_parent">
  <img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/>
</a>

In [1]:
import math
import matplotlib.pyplot as plt
import numpy as np
import scipy.integrate
from numpy.polynomial.polynomial import Polynomial
import math

Deklarujemy "całkowaną" przez nas funkcję

In [2]:
def func(x_):
    return 6*x_*math.exp(0.5*x_)+1.5*x_**2*math.exp(0.5*x_)


In [4]:
def newton_cotes(func, a, b, n):
    # dictionary that includes sigma and ns values for certain n
    coeffs = {
        1: {
            "sigma": [1, 1],
            "ns": 2,
        },
        2: {
            "sigma": [1, 4, 1],
            "ns": 6,
        },
        3: {
            "sigma": [1, 3, 3, 1],
            "ns": 8,
        },
        4: {
            "sigma": [7, 32, 12, 32, 7],
            "ns": 90,
        },
        5: {
            "sigma": [19, 75, 50, 50, 75],
            "ns": 288,
        },
    }

    integral = 0
    h = (b - a) / n
    # iterate 
    for i in range(n + 1):
        # increment certain sigma muliplied by certain function val
        integral += coeffs[n]["sigma"][i] * func(a + i * h)
    integral *= (b - a) / coeffs[n]["ns"]
    return integral


def newton_cotes_trapezoid(func, a, b, n):
    h = (b - a) / n
    integral = func(a) + func(b)  # find the func(a) and func(b)

    for i in range(1, n):
        integral += 2 * (func(a + i * h))
    integral *= h / 2.0
    return integral


def newton_cotes_simpson(func, a, b, n):
    h = (b - a) / n
    integral = func(a) + func(b)

    for i in range(1, n):
        # even
        if i % 2 == 0:
            integral += 2 * (func(a + i * h))
        # uneven 
        else:
            integral += 4 * (func(a + i * h))
    integral *= h / 3.0
    return integral


Przetestujmy wszystkie powyższe funkcje numeryczne, a mianowicie kolejno podstawową, trapezową i Simpsona

In [5]:
newton_cotes(func, 1, 2, 4)

27.673223590464858

In [6]:
newton_cotes_trapezoid(func, 1, 2, 4)

27.860921634657014

In [7]:
newton_cotes_simpson(func, 1, 2, 4)

27.67370944173623

Sprawdźmy jak równanie będzie wyglądać w postaci symbolicznej:

In [8]:
import sympy as sym

x = sym.symbols("x")

def newton_cotes_sym(a, b):
    integral = sym.integrate(6*x*sym.exp(0.5*x)+1.5*x**2*sym.exp(0.5*x), (x, a, b))
    return integral


In [9]:
def integrated_func(x_):
    return 3.0*x_**2*math.exp(0.5*x_)

Wywołajmy funkcję symboliczną, która obliczy nam symbolicznie całkę 

In [10]:
newton_cotes_sym(1,2)

27.6732181294082

Funkcja mierząca błąd

In [11]:
def error(value):
    return abs(newton_cotes_sym(1,2) - value)

In [12]:
print(f"{newton_cotes_sym(1,2)}\t{error(newton_cotes_sym(1,2))}")
print(f"{newton_cotes(func, 1, 2, 4)}\t{error(newton_cotes(func, 1, 2, 4))}")
print(f"{newton_cotes_trapezoid(func, 1, 2, 4)}\t{error(newton_cotes_trapezoid(func, 1, 2, 4))}")
print(f"{newton_cotes_simpson(func, 1, 2, 4)}\t{error(newton_cotes_simpson(func, 1, 2, 4))}")

27.6732181294082	0
27.673223590464858	0.00000546105669840813
27.860921634657014	0.187703505248855
27.67370944173623	0.000491312328069426


Jak widzimy spośród tych metod podstawowa metoda Newtona-Cotes'a jest najbardziej dokładna


| Metoda | Wynik | Błąd |
| --- | --- | --- |
| Symboliczna | 27.6732181294082 | 0 |
| Podstawowa | 27.673223590464858 | .0.00000546105669840813 |
| Trapez | 27.860921634657014 | .0.187703505248855 |
| Simpsona | 27.67370944173623 | .0.000491312328069426 |

	
	
	